### DISCUSSION TITLE
Window layering only applies to the group of the window on X11

### DISCUSSION THREAD

**Discussion by nathannino** [qtile/qtile#4913] [Q&A] [answered]
Title...

I've been trying to set the window layering myself using hooks due to not liking qtile's default behaviour. (Mostly the part where floating layering order is based on when they were moved to floating, and putting them back into tilling mode doesn't put them behind windows that were floating before the window now in tilling was set to floating. Yeah, there is a way to bring the window to the front, and by default it's middle click, but I would rather have that done automatically)

Now, I've tested my config on both the wayland core and the x11 core, and I have noticed that on the x11 core, the layering seems to be based on the group. If a window is floating in between 2 screens, and a second, tilling window is on the group of the screen I want to move the floating window to, I need to move it around the halfway point into the second screen for it to actually move up, despite the fact that I am calling `move_to_top()` on every floating window after calling it on every tilling window. I have tested it on the wayland core, and my automatic layering config seems to work without issues there, but since I am also getting graphical glitches on the wayland core, that's not a viable solution.

**Is there a way to set the layering of the window on the x11 core so that a window can be set as being above another window even if they are on different groups?**

Theses are excerpts of my current config :
```python
def _reorder_window_zaxis(window_list) :
    for window_index in range(len(window_list)) :
        window_list[window_index].keep_above(False) # Maybe this will work?
        window_list[window_index].move_to_top()


def reorder_window_zaxis():
    global window_tiling
    global window_floating
    global window_fullscreen
    _reorder_window_zaxis(window_tiling)
    _reorder_window_zaxis(window_floating)
    _reorder_window_zaxis(window_fullscreen)
```
As for the order inside of each list :
```python
def deregister_window(window_killed) :
    global window_tiling
    global window_floating
    global window_fullscreen
    list_silence_remove(window_tiling,window_killed)
    list_silence_remove(window_floating,window_killed)
    list_silence_remove(window_fullscreen,window_killed)


def mark_as_latest_floating(window_obj) :
    global window_floating
    deregister_window(window_obj)
    window_floating.append(window_obj)


def move_window_state(window_to_update,deregister=True,check_exists=False) :
    global window_tiling
    global window_floating
    global window_fullscreen
    window_info = window_to_update.info()

    # Exit if already in the correct spot
    if (check_exists) :
        if (window_info.get("fullscreen",False) and window_to_update in window_fullscreen) :
            return
        elif (window_info.get("floating",False) and window_to_update in window_floating) :
            return
        elif (((not window_info.get("fullscreen",False)) and (not window_info.get("floating",False))) and window_to_update in window_tiling) :
            return

    # Remove if needed
    if (deregister) :
        deregister_window(window_to_update)

    # Put inside correct spot
    if (window_info.get("group") is None) :
        return # TODO : This is bar. We need to keep bar in mind or something

    if (window_info.get("fullscreen", False)) :
        window_fullscreen.append(window_to_update)
    elif (window_info.get("floating", False)) :
        window_floating.append(window_to_update)
    else :
        window_tiling.append(window_to_update)
```
With `move_window_state()`, `deregister_window()`, `mark_as_latest_floating()` and `reorder_window_zaxis()` being called by hooks and keybinds

Finally, the wayland graphical glitches are probably nvidia related tbh

**Comment by elParaguayo** [comment, marked as answer]
That behaviour is intentional. The reasoning was, if we didn't do stacking based on windows in a group, commands like `move_up` may have no obvious effect if a window is only moved higher in relation to a window in a different group.

I'm open to suggestions for improvements.

**Reply by nathannino** [reply]
hm, so in that case, the fact it worked in wayland would be a bug then.

> commands like move_up may have no obvious effect if a window is only moved higher in relation to a window in a different group.

Yeah, thinking about it from setting the command to a keybinding or using `qtile cmd-obj`, that does make sense.


> I'm open to suggestions for improvements.

I'll be honest, considering this is the intended behavior, I don't really know what you could do, and I don't think I am familiar enough with qtile's internals to know what a solution would require. Even I know that as much as saying to support both use cases is easy, actually implementing it and maintaining it would be much harder. 

However, in case it might help, I've disabled my own `reorder_window_zaxis()` and set both `floats_kept_above` and `bring_front_click` to true in order to note my pain points with how it's currently working :

In the end, I found two things that caused me issues. First of all, if I have at least two floating windows in the same group and no tilling windows, and I tell the window at the top of the layer to tile itself, that window will end up covering up the other floating windows, as it will not go back down. If I did have tiling windows in the group, then it works as expected... that might actually be a bug too, now that I think about it. I just never tested it with a tiling window in the group before writing this.

Second of all, if I am attempting to move a floating window from one screen to the other, what ends up happening is that if I don't grab it with my mouse cursor on the side farther from the screen I want to move it to and that the screen I want to move it to has a tiling window, I will end up with two floating windows with the window I wanted to move stuck in between the two screens, as `follow_mouse_focus` will move the focus to the window with the higher layer, which will mean that the drag keybinding will stop applying to the window I wanted to move and start applying to the new focused window. 
Now, this doesn't happen all of the times, as it depends on whether the window I wanted to move is higher than the window on the other screen/group, but it is disruptive when it happens, as I have now have two windows in the wrong spot instead of one.

If there wasn't any cost (time/complexity/money/etc) to implementing it, I think what I would expect is that whatever window I am moving should be at the front at all times, and once released, that window should be set in front of every group it has part of it's geometry inside of. However, since `move_up()` is a thing, it should also be a toggleable option, like `floats_kept_above`. That sounds really hard to do however.

I hope at least part of this is helpful, and thanks for answering my question

**Reply by elParaguayo** [reply]
> In the end, I found two things that caused me issues. First of all, if I have at least two floating windows in the same group and no tilling windows, and I tell the window at the top of the layer to tile itself, that window will end up covering up the other floating windows, as it will not go back down. If I did have tiling windows in the group, then it works as expected... that might actually be a bug too, now that I think about it. I just never tested it with a tiling window in the group before writing this.

That is not expected behaviour so, yes, it's a bug.
 
> Second of all, if I am attempting to move a floating window from one screen to the other, what ends up happening is that if I don't grab it with my mouse cursor on the side farther from the screen I want to move it to and that the screen I want to move it to has a tiling window, I will end up with two floating windows with the window I wanted to move stuck in between the two screens, as `follow_mouse_focus` will move the focus to the window with the higher layer, which will mean that the drag keybinding will stop applying to the window I wanted to move and start applying to the new focused window. Now, this doesn't happen all of the times, as it depends on whether the window I wanted to move is higher than the window on the other screen/group, but it is disruptive when it happens, as I have now have two windows in the wrong spot instead of one.

> If there wasn't any cost (time/complexity/money/etc) to implementing it, I think what I would expect is that whatever window I am moving should be at the front at all times, and once released, that window should be set in front of every group it has part of it's geometry inside of. However, since `move_up()` is a thing, it should also be a toggleable option, like `floats_kept_above`. That sounds really hard to do however.

Yeah, that does sound annoying. Your solution sounds like the route I'd probably try,
